File: Week3 Assignment/nth_node_end.c
//Nth node from end

#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

void insertAtEnd(struct Node** head, int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;

    if (*head == NULL) {
        *head = newNode;
        return;
    }

    struct Node* temp = *head;
    while (temp->next != NULL)
        temp = temp->next;
    temp->next = newNode;
}

void nthFromEnd(struct Node* head, int n) {
    struct Node* fast = head;
    struct Node* slow = head;
    int i;
    for (i = 0; i < n; i++) {
        if (fast == NULL) {
            printf("List has fewer than %d nodes\n", n);
            return;
        }
        fast = fast->next;
    }

    while (fast != NULL) {
        slow = slow->next;
        fast = fast->next;
    }

    printf("%dth node from end is: %d\n", n, slow->data);
}

int main() {
    struct Node* head = NULL;

    insertAtEnd(&head, 10);
    insertAtEnd(&head, 20);
    insertAtEnd(&head, 30);
    insertAtEnd(&head, 40);
    insertAtEnd(&head, 50);

    nthFromEnd(head, 2); // 2nd node from end ? 40

    return 0;
}

File: Week3 Assignment/nth_node_end.py
#Nth node from end

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insert_at_end(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node

    def nth_from_end(self, n):
        fast = self.head
        slow = self.head

        for _ in range(n):
            if not fast:
                print(f"List has fewer than {n} nodes")
                return
            fast = fast.next

        while fast:
            slow = slow.next
            fast = fast.next

        print(f"{n}th node from end is: {slow.data}")

# Example usage
ll = LinkedList()
ll.insert_at_end(10)
ll.insert_at_end(20)
ll.insert_at_end(30)
ll.insert_at_end(40)
ll.insert_at_end(50)

ll.nth_from_end(2)  # 40
